### PR TITLE
Enhancement separate smap index

### PR DESCRIPTION
**Pull request by simitt**
This PR is based on PR #578.

### CODE REVIEW

File: CHANGELOG.asciidoc [hunk 1]
```diff
@@ -43,6 +44,7 @@ https://github.com/elastic/apm-server/compare/71df0d96445df35afe27f38bcf734a0828
 - Add `process.ppid` as optional field {pull}564[564]
 - Change `error.culprit` after successfully applying sourcemapping {pull}520[520]
 - Make `transaction.name` optional {pull}554[554]
+- Change generationg of config files and supported options {pull}578[578]
```
Comment (graphaelli) [inline CHANGELOG.asciidoc]: typo

**Comment by graphaelli** [comment]
Is this really a breaking change?  Defaults in 6.2 would put sourcemaps in `apm-*`, this puts sourcemaps in `apm-sourcemap*`.  If a user used this and then downgraded back to 6.2, apm-server would continue to look for sourcemaps in `apm-*` and would find them, even if they were in `apm-sourcemap*`.  Maybe moot since this feature is beta.

File: sourcemap/elasticsearch.go [hunk 2]
```diff
@@ -82,9 +82,9 @@ func query(id Id) map[string]interface{} {
 			"bool": map[string]interface{}{
 				"must": []map[string]interface{}{
 					{"term": map[string]interface{}{"processor.name": "sourcemap"}},
-					{"term": map[string]interface{}{"sourcemap.bundle_filepath": id.Path}},
-					{"term": map[string]interface{}{"sourcemap.service.name": id.ServiceName}},
-					{"term": map[string]interface{}{"sourcemap.service.version": id.ServiceVersion}},
+					{"match": map[string]interface{}{"sourcemap.bundle_filepath": id.Path}},
```
[the remark this answers is not in the document]
Comment (simitt) [reply sourcemap/elasticsearch.go]: A`term` query only works on keywords. When supporting custom index names for sourcemap no ES template is (automatically) loaded for this index, indexing a`text` and `keyword` entry. With `match` we can cover this.

**Comment by simitt** [comment]
@graphaelli you are right, it is not a breaking change (I wrote this when I started out with using a `sourcemap` index, but then figured, prefixing it with `apm` might be a better idea) - will change.

File: tests/system/apmserver.py [hunk 3]
```diff
@@ -303,11 +303,25 @@ def config(self):
...
+                    "smap_index": "sourcemap"})
+        return cfg
+
+
 class ExpvarBaseTest(ServerBaseTest):
     config_overrides = {}
 
     def config(self):
-        cfg = super(ServerBaseTest, self).config()
+        cfg = super(ExpvarBaseTest, self).config()
```
Comment (jalvz) [inline tests/system/apmserver.py]: good eyes :)

File: sourcemap/elasticsearch.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (roncohen) [reply sourcemap/elasticsearch.go]: does that mean the sourcemap itself is now going to get indexed?

**Comment by simitt** [comment]
Checking indexing again and reopen PR afterwards.

**Comment by simitt** [comment]
@elastic/apm-server  after our discussions today, I have changed the sourcemap index to default to `apm-{beat.version}-sourcemap`, and the query index pattern to `apm-*-sourcemap`.

**Comment by jalvz** [comment]
is possible to have this independent from 578 (for easier review)?

**Comment by simitt** [comment]
@jalvz after PR #578 was merged, this PR is cleaner now - please review.

File: beater/config.go [hunk 4]
```diff
@@ -83,10 +87,12 @@ func (c *FrontendConfig) SmapMapper() (sourcemap.Mapper, error) {
 	if smap.mapper != nil {
 		return c.SourceMapping.mapper, nil
 	}
+	re := regexp.MustCompile("%.*{.*beat.version.?}")
```
Comment (jalvz) [inline beater/config.go]: I think this method should be documented stating that it modifies its object, and that if a second time is called, nothing happens, ie. if i got it right this behaves like this:
```
c.beatVersion = 1
c.SmapMapper()
print  c.SourceMapping.index  // apm-1

c.beatVersion = 2
c.SmapMapper()
print  c.SourceMapping.index  // still apm-1
```

**Comment by jalvz** [comment]
I think this needs a test somewhere that POSTes a sourcemap and eg a transaction, and then asserts that the sourcemap and the transactions are indeed in different indices.

File: beater/config.go (hunk 4 above)
Comment (simitt) [reply beater/config.go]: I don't think that the `beat.version` can change (from a logical perspective) without restarting the server.

**Comment by simitt** [comment]
Could you take a look at this test https://github.com/elastic/apm-server/blob/master/tests/system/test_integration.py#L287.

First a `sourcemap` is uploaded and verified that it is there, by using the [sourcemap index](https://github.com/simitt/apm-server/blob/25136bb16f0a2640ca85e4b47b43164f8093f70c/tests/system/apmserver.py#L253) which is [specifically set](https://github.com/simitt/apm-server/blob/25136bb16f0a2640ca85e4b47b43164f8093f70c/tests/system/apmserver.py#L312) and [set in the config](https://github.com/simitt/apm-server/blob/25136bb16f0a2640ca85e4b47b43164f8093f70c/tests/system/config/apm-server.yml.j2#L65).
Then a `error` is [uploaded and verified](https://github.com/simitt/apm-server/blob/25136bb16f0a2640ca85e4b47b43164f8093f70c/tests/system/apmserver.py#L171)

As the `index` and the `smap_index` are defined to be different, this tests that the docs are in different indices. Please let me know if I am missing something. else here

File: beater/config.go (hunk 4 above)
Comment (jalvz) [reply beater/config.go]: But the point is the same. I don't say it's a bug, just that I would appreciate documentation when any kind of side effects are going on; or rename it to make it evident (eg `SetSourceMapperIfNil`)

**Comment by jalvz** [comment]
that works, thanks

File: beater/config.go (hunk 4 above)
Comment (simitt) [reply beater/config.go]: that's true, will change.

Comment (simitt) [reply beater/config.go]: @jalvz I did refactor a bit, please have a look.

As this PR doesn't introduce the problem, I suggest to create an Issue and address once we work on config reloading.

Good finding.

Comment (jalvz) [reply beater/config.go]: LGTM. an issue is not necessary for me. It wasn't a comment on config reloading, but on code explicitness. thanks.

**Pull request merged**